src/OrchardCore.Modules/OrchardCore.Media/SecureMediaPermissions.cs

```diff
@@ -0,0 +1,168 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using OrchardCore.Environment.Cache;
+using OrchardCore.Media.Services;
+using OrchardCore.Modules;
+using OrchardCore.Security.Permissions;
+
+namespace OrchardCore.Media
+{
+    [Feature("OrchardCore.Media.Security")]
+    public class SecureMediaPermissions : IPermissionProvider
+    {
+        // Note: The ManageMediaFolder permission grants all access, so viewing must be implied by it too.
+        public static readonly Permission ViewMedia = new("ViewMediaContent", "View media content in all folders", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewRootMedia = new("ViewRootMediaContent", "View media content in the root folder", new[] { ViewMedia });
+        public static readonly Permission ViewOthersMedia = new("ViewOthersMediaContent", "View others media content", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewOwnMedia = new("ViewOwnMediaContent", "View own media content", new[] { ViewOthersMedia });
+
+        private static readonly Permission _viewMediaTemplate = new("ViewMediaContent_{0}", "View media content in folder '{0}'", new[] { ViewMedia });
```

Comment:
I think you need to make this public like the others, that's why the order is matter :)

Reply:
I don't think so. One can argue that `PermissionTemplates` can be private though...

Reply:
Agree if you are not using it as permission, but you declare it as `Permission`

Reply:
Really everything should be private except what you need to do authorization calls outside.

Reply:
I changed the visibility, but note that this may have to be reversed in future. Some other code may want to reuse the dynamic permissions to check a requirement of a folder for a different purpose.